tests/prague/eip6206_jumpf/contracts.py

```diff
@@ -0,0 +1,244 @@
+    )
+
+
+def container_name(c: Container):
+    """
+    Return the name of the container for use in pytest ids.
+    """
+    if hasattr(c, "name"):
+        return c.name
+    else:
+        return c.__class__.__name__
+
+
+def generate_jumpf_target_rules():
+    """
+    Generate tests for JUMPF where we are testing the validity of the JUNMPF target.
+    We are not testing stack so a lot of the logic is to get correct stack values.
+    """
+    valid = []
+    invalid = []
+    for current_outputs in [NON_RETURNING_SECTION, 0, 2, 4]:
+        current_non_returning = current_outputs == NON_RETURNING_SECTION
+        current_height = 0 if current_non_returning else current_outputs
+        for target_outputs in [NON_RETURNING_SECTION, 0, 2, 4]:
+            target_non_returning = target_outputs == NON_RETURNING_SECTION
+            target_height = 0 if target_non_returning else target_outputs
+            delta = (
+                0
+                if target_non_returning or current_non_returning
+                else target_outputs - current_height
+            )
+            current_extra_push = max(0, current_height - target_height)
+            current_section = Section.Code(
+                code=Op.PUSH0 * (current_height)
+                + Op.CALLDATALOAD(0)
+                + Op.RJUMPI[1]
+                + (Op.STOP if current_non_returning else Op.RETF)
+                + Op.PUSH0 * current_extra_push
+                + Op.JUMPF[2],
+                code_inputs=0,
+                code_outputs=current_outputs,
+                max_stack_height=current_height + max(1, current_extra_push),
+            )
+            target_section = Section.Code(
+                code=((Op.PUSH0 * delta) if delta >= 0 else (Op.POP * -delta))
+                + Op.CALLF[3]
+                + (Op.STOP if target_non_returning else Op.RETF),
+                code_inputs=current_height,
+                code_outputs=target_outputs,
+                max_stack_height=max(current_height, current_height + delta),
+            )
+
+            container = Container(
+                name="target_co-%s_to-%s"
+                % (
+                    "N" if current_non_returning else current_outputs,
+                    "N" if target_non_returning else target_outputs,
+                ),
+                sections=[
+                    quick_code(Op.JUMPF[1], height=0 if current_non_returning else current_height)
+                    if current_non_returning
+                    else quick_code(
+                        Op.CALLF[1](0, 0) + Op.STOP,
+                        height=0 if current_non_returning else 2 + current_outputs,
+                    ),
+                    current_section,
+                    target_section,
+                    quick_code(Op.SSTORE(0, 1) + Op.RETF, outputs=0, height=2),
+                ],
+            )
+
+            # now sort validity...
+            if target_non_returning:
```

Comment:
I think this function logic is too hidden from the actual test. 
we can test exactly this function and use parametrisation to hide the loops 
then just yell more eof_test functions

Reply:
When I parameterized the tests now I check at the end against validity rules and add exceptions.